components/date-picker/date-picker.jsx

```diff
@@ -509,69 +560,32 @@ class Datepicker extends React.Component {
 			this.props.assistiveText
 		);
 
-		const clonedInputProps = {
-			disabled:
-				(this.props.children && !!this.props.children.props.disabled) ||
-				this.props.disabled,
-			iconRight: (this.props.children &&
-				!!this.props.children.props.iconRight) || (
-				<InputIcon
-					// Remove || for assistiveText at next breaking change
-					assistiveText={{
-						icon:
-							this.props.assistiveTextOpenCalendar ||
-							assistiveText.openCalendar, // eslint-disable-line react/prop-types
-					}}
-					aria-haspopup
-					aria-expanded={this.getIsOpen()}
-					category="utility"
-					name="event"
-					onClick={this.openDialog}
-					type="button"
-				/>
-			),
-			id: this.getId(),
-			inputRef: (component) => {
-				this.setInputRef(component);
-			},
-			label:
-				(this.props.children && this.props.children.props.label) ||
-				this.props.label || // eslint-disable-line react/prop-types
-				labels.label,
-			onBlur:
-				(this.props.children && this.props.children.props.onBlur) ||
-				this.props.onBlur, // eslint-disable-line react/prop-types
-			onChange: this.handleInputChange,
-			onClick: () => {
-				this.openDialog();
-				if (this.props.children && this.props.children.props.onClick) {
-					this.props.children.props.onClick();
-				}
-			},
-			onFocus:
-				(this.props.children && this.props.children.props.onFocus) ||
-				this.props.onFocus, // eslint-disable-line react/prop-types
-			onKeyDown:
-				(this.props.children && this.props.children.props.onKeyDown) ||
-				this.handleKeyDown,
-			placeholder:
-				(this.props.children && this.props.children.props.placeholder) ||
-				this.props.placeholder || // eslint-disable-line react/prop-types
-				labels.placeholder,
-			required:
-				(this.props.children && this.props.children.props.required) ||
-				this.props.required, // eslint-disable-line react/prop-types
-			value:
-				(this.props.children && this.props.children.props.value) ||
-				this.state.inputValue,
+		/**
+		 * 1. DEFAULT: Use default props or state if present.
+		 * 2. DEPRECATED API: Use old "first-level" props that have been deprecated.
+		 * 3. DEPRECATED API: If `children` is present, use props from single child which should be an `<Input/>`
+		 * 4. CURRENT API: Use composition with props spread merge from `input` prop.
+		 * */
+		const finalInputProps = {
+			...this.getDefaultInputProps({ assistiveText }),
+			...this.getTopLevelDeprecatedComponentProps({
```

Comment:
Same comment on this one.  I think maybe it might be cleaner to do all of this prop merging in one place.  Is this the best place to put this object creation?  Or should all of these methods just be inlined into one method called `getInputProps`?  Then the "logic" of getting the props for the input would be contained in one obvious method definition (albeit a long one) instead of spread out in 4 different methods across this file.  I think you can make a single method more readable with judicious use of clarifying variables inside of it rather than how this is currently written.